README.md
# exif2css

[![npm version](https://badge.fury.io/js/exif2css.svg)](https://npmjs.org/package/exif2css)

`exif2css` Converts EXIF Orientation To CSS Transform Rules (in 772 bytes gzipped).

```sh
yarn add -E exif2css
```

![exif2css](exif2css.jpg)

## Table Of Contents

- [Table Of Contents](#table-of-contents)
- [API](#api)
- [`exif2css(orientation: (number|string)): Exif2CssReturn`](#exif2cssorientation-numberstring-exif2cssreturn)
  * [`Exif2CssReturn`](#type-exif2cssreturn)
- [Usage](#usage)
  * [As Node Module](#as-node-module)
  * [As Script](#as-script)
- [Testing](#testing)
- [Copyright](#copyright)

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/0.svg?sanitize=true"></a></p>

## API

The package is available by importing its default function:

```js
import exif2css from 'exif2css'
```

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/1.svg?sanitize=true"></a></p>

## `exif2css(`<br/>&nbsp;&nbsp;`orientation: (number|string),`<br/>`): Exif2CssReturn`

Converts an integer or a string representing EXIF orientation into required CSS with _transfrom_ and optionally _transform-origin_ properties. They can then be used as needed, possibly prefixing the rules with browser-specific tags (e.g., `-webkit-transform` and `-webkit-transform-origin`).

One known issue is that with orientations > 4, the transformed image will have different dimensions from its box, so that whitespace might appear on the right and at the bottom of the image.

```js
import exif2css from 'exif2css'

['not-an-exif-orientation', 1,2,3,4,5,6,7,8]
  .forEach((i) => {
    console.log('Orientation: %s', i)
    const result = exif2css(i)
    console.log(result)
    console.log()
  })
```
```js
Orientation: not-an-exif-orientation
{}

Orientation: 1
{}

Orientation: 2
{ transform: 'rotateY(180deg)',
  transforms: { rotateY: 180 },
  transformStrings: { rotateY: 'rotateY(180deg)' } }

Orientation: 3
{ transform: 'rotate(180deg)',
  transforms: { rotate: 180 },
  transformStrings: { rotate: 'rotate(180deg)' } }

Orientation: 4
{ transform: 'rotate(180deg) rotateY(180deg)',
  transforms: { rotate: 180, rotateY: 180 },
  transformStrings: { rotate: 'rotate(180deg)', rotateY: 'rotateY(180deg)' } }

Orientation: 5
{ transform: 'rotate(270deg) rotateY(180deg)',
  'transform-origin': 'top left',
  transforms: { rotate: 270, rotateY: 180 },
  transformStrings: { rotate: 'rotate(270deg)', rotateY: 'rotateY(180deg)' } }

Orientation: 6
{ transform: 'translateY(-100%) rotate(90deg)',
  'transform-origin': 'bottom left',
  transforms: { translateY: -1, rotate: 90 },
  transformStrings: { translateY: 'translateY(-100%)', rotate: 'rotate(90deg)' } }

Orientation: 7
{ transform: 'translateY(-100%) translateX(-100%) rotate(90deg) rotateY(180deg)',
  'transform-origin': 'bottom right',
  transforms: { translateY: -1, translateX: -1, rotate: 90, rotateY: 180 },
  transformStrings: 
   { translateY: 'translateY(-100%)',
     translateX: 'translateX(-100%)',
     rotate: 'rotate(90deg)',
     rotateY: 'rotateY(180deg)' } }

Orientation: 8
{ transform: 'translateX(-100%) rotate(270deg)',
  'transform-origin': 'top right',
  transforms: { translateX: -1, rotate: 270 },
  transformStrings: { translateX: 'translateX(-100%)', rotate: 'rotate(270deg)' } }
```

__<a name="type-exif2cssreturn">`Exif2CssReturn`</a>__: The return type of the function.

|       Name       |                                    Type                                     |                                                      Description                                                       |
| ---------------- | --------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| transform        | _string_                                                                    | The complete CSS `transform` rule that contains all transforms.                                                        |
| transform-origin | _('top left'\|'top right'\|'bottom left'\|'bottom right')_                  | The transform origin CSS rule for orientations >= 5.                                                                   |
| transforms       | _{translateY: number, translateX: number, rotate: number, rotateY: number}_ | The raw transforms as numbers, where translates are either `-1` or `1` and rotations are either `90`, `180` and `270`. |
| transformStrings | _{translateY: string, translateX: string, rotate: string, rotateY: string}_ | The transforms split by individual rules that can be applied in the browser.                                           |

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/2.svg?sanitize=true"></a></p>

## Usage

The module can be either required in _Node.JS_, or downloaded as the compiled file from the _dist_ folder and inserted on the webpage.

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/3.svg?sanitize=true" width="15"></a></p>

### As Node Module

The package is published both as CommonJS module with the `main` field, and as a ES6 module with the `module` field. Node will automatically pick up the `main` version, whereas some bundles will be able to use the module.

```bash
yarn add -E exif2css
npm i exif2css
```

```js
import exif2css from 'exif2css' // or
const exif2css = require('exif2css')

const css = exif2css(6)
```

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/4.svg?sanitize=true" width="15"></a></p>

### As Script

_Exif2Css_ has been compiled with [_Depack_](https://github.com/dpck/depack) using Google Closure Compiler. Download the [file](https://github.com/demimonde/exif2css/blob/master/dist/exif2css.js) manually and embed it on the webpage.

```html
<img src="some-image.jpg">
<script src="exif2css.js"></script>
<script>
  var img = document.querySelector('img')
  var orientation = 6
  var css = exif2css(orientation)

  if (css.transform) {
      img.style.transform = css.transform
  }
  if (css['transform-origin']) {
      img.style['transform-origin'] = css['transform-origin']
  }
</script>
```

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/5.svg?sanitize=true"></a></p>

## Testing

The module has been automatically tested in Chrome by inserting pre-compiled images with set orientation, applying generated styles from `exif2css`, taking screenshots and comparing them against the expected image, therefore everything works properly.

![Testing exif2css diff](artifacts/1_diff.png)

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/6.svg?sanitize=true"></a></p>

## Copyright

(c) [Demimonde][1] 2019

[1]: https://demimonde.cc

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/-1.svg?sanitize=true"></a></p>

CHANGELOG.md
## 20 March 2019

### [1.3.0](https://github.com/demimonde/exif2css/compare/v1.2.0...v1.3.0)

- [package] Rewrite the package and tests; move to _Demimonde_ GitHub org; build with _Depack_.
- [doc] Add some JSDoc, use _Documentary_.

## 19 March 2019

### 1.0.0

- Create `exif2css` with [`mnp`][https://mnpjs.org]
- [repository]: `src`, `test`


test/spec/unit.js
const exif2css = require('../../src/index')
const assert = require('assert')

module.exports = {
    'should return empty object for non-exif value': () => {
        const res = exif2css('non-exif')
        assert(JSON.stringify(res) === JSON.stringify({}))
    },
    '1': {
        'empty object': () => {
            const res = exif2css(1)
            assert(JSON.stringify(res) === JSON.stringify({}))
        },
    },
    '2': {
        'transform: rotateY(180deg)': () => {
            const res = exif2css(2)
            assert(res.transform === 'rotateY(180deg)')
        },
        transforms: {
            'rotateY === 180': () => {
                const res = exif2css(2)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transforms.rotateY === 180)
            },
        },
        transformStrings: {
            'rotateY === "rotateY(180deg)"': () => {
                const res = exif2css(2)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transformStrings.rotateY === 'rotateY(180deg)')
            },
        },
    },
    '3': {
        'transform: rotate(180deg)': () => {
            const res = exif2css(3)
            assert(res.transform === 'rotate(180deg)')
        },
        transforms: {
            'rotate === 180': () => {
                const res = exif2css(3)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 180)
            },
        },
        transformStrings: {
            'rotate === "rotate(180deg)"': () => {
                const res = exif2css(3)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(180deg)')
            },
        },
    },
    '4': {
        'transform: rotate(180deg) rotateY(180deg)': () => {
            const res = exif2css(4)
            assert(res.transform === 'rotate(180deg) rotateY(180deg)')
        },
        transforms: {
            'rotate === 180': () => {
                const res = exif2css(4)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 180)
            },
            'rotateY === 180': () => {
                const res = exif2css(4)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transforms.rotateY === 180)
            },
        },
        transformStrings: {
            'rotate === "rotate(180deg)"': () => {
                const res = exif2css(4)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(180deg)')
            },
            'rotateY === "rotateY(180deg)"': () => {
                const res = exif2css(4)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transformStrings.rotateY === 'rotateY(180deg)')
            },
        },
    },
    '5': {
        'transform: rotate(270deg) rotateY(180deg)': () => {
            const res = exif2css(5)
            assert(res.transform === 'rotate(270deg) rotateY(180deg)')
        },
        'transform-origin: top left': () => {
            const res = exif2css(5)
            assert(res['transform-origin'] === 'top left')
        },
        transforms: {
            'rotate === 270': () => {
                const res = exif2css(5)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 270)
            },
            'rotateY === 180': () => {
                const res = exif2css(5)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transforms.rotateY === 180)
            },
        },
        transformStrings: {
            'rotate === "rotate(270deg)"': () => {
                const res = exif2css(5)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(270deg)')
            },
            'rotateY === "rotateY(180deg)"': () => {
                const res = exif2css(5)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transformStrings.rotateY === 'rotateY(180deg)')
            },
        },
    },
    '6': {
        'transform: translateY(-100%) rotate(90deg)': () => {
            const res = exif2css(6)
            assert(res.transform === 'translateY(-100%) rotate(90deg)')
        },
        'transform-origin: bottom left': () => {
            const res = exif2css(6)
            assert(res['transform-origin'] === 'bottom left')
        },
        transforms: {
            'translateY === -1': () => {
                const res = exif2css(6)
                assert(res.transforms !== undefined)
                assert('translateY' in res.transforms)
                assert(res.transforms.translateY === -1)
            },
            'rotate === 90': () => {
                const res = exif2css(6)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 90)
            },
        },
        transformStrings: {
            'translateY === "translateY(-100%)"': () => {
                const res = exif2css(6)
                assert(res.transforms !== undefined)
                assert('translateY' in res.transforms)
                assert(res.transformStrings.translateY === 'translateY(-100%)')
            },
            'rotate === "rotate(90deg)"': () => {
                const res = exif2css(6)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(90deg)')
            },
        },
    },
    '7': {
        'transform: translateY(-100%) translateX(-100%) rotate(90deg) rotateY(180deg)': () => {
            const res = exif2css(7)
            assert(res.transform === 'translateY(-100%) translateX(-100%) rotate(90deg) rotateY(180deg)')
        },
        'transform-origin: bottom right': () => {
            const res = exif2css(7)
            assert(res['transform-origin'] === 'bottom right')
        },
        transforms: {
            'translateY === -1': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('translateY' in res.transforms)
                assert(res.transforms.translateY === -1)
            },
            'translateX === -1': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('translateX' in res.transforms)
                assert(res.transforms.translateX === -1)
            },
            'rotate === 90': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 90)
            },
            'rotateY === 180': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transforms.rotateY === 180)
            },
        },
        transformStrings: {
            'translateY === "translateY(-100%)"': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('translateY' in res.transforms)
                assert(res.transformStrings.translateY === 'translateY(-100%)')
            },
            'translateX === "translateX(-100%)"': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('translateX' in res.transforms)
                assert(res.transformStrings.translateX === 'translateX(-100%)')
            },
            'rotate === "rotate(90deg)"': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(90deg)')
            },
            'rotateY === "rotateY(180deg)"': () => {
                const res = exif2css(7)
                assert(res.transforms !== undefined)
                assert('rotateY' in res.transforms)
                assert(res.transformStrings.rotateY === 'rotateY(180deg)')
            },
        },
    },
    '8': {
        'transform: translateX(-100%) rotate(270deg)': () => {
            const res = exif2css(8)
            assert(res.transform === 'translateX(-100%) rotate(270deg)')
        },
        'transform-origin: top right': () => {
            const res = exif2css(8)
            assert(res['transform-origin'] === 'top right')
        },
        transforms: {
            'translateX === -1': () => {
                const res = exif2css(8)
                assert(res.transforms !== undefined)
                assert('translateX' in res.transforms)
                assert(res.transforms.translateX === -1)
            },
            'rotate === 270': () => {
                const res = exif2css(8)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transforms.rotate === 270)
            },
        },
        transformStrings: {
            'translateX === "translateX(-100%)"': () => {
                const res = exif2css(8)
                assert(res.transforms !== undefined)
                assert('translateX' in res.transforms)
                assert(res.transformStrings.translateX === 'translateX(-100%)')
            },
            'rotate === "rotate(270deg)"': () => {
                const res = exif2css(8)
                assert(res.transforms !== undefined)
                assert('rotate' in res.transforms)
                assert(res.transformStrings.rotate === 'rotate(270deg)')
            },
        },
    },
}


test/mask/index.js
import { makeTestSuite } from 'zoroaster'
import Context from '../context'
import exif2css from '../../src'

const ts = makeTestSuite('test/result', {
  async getResults(input) {
    const res = await exif2css({
      text: input,
    })
    return res
  },
  context: Context,
})

// export default ts


test/spec/index.js
const fs = require('fs')
const resemble = require('node-resemble-js')
const path = require('path')
const assert = require('assert')
const webdriver = require('selenium-webdriver')
const phantomjs = require('phantomjs-prebuilt').path
const browserify = require('browserify')
const index = require('../../src/index')

const customPhantom = webdriver.Capabilities.phantomjs()
customPhantom.set("phantomjs.binary.path", phantomjs)

let driver
let bundle

const expectedImagePath = path.join(__dirname, '../fixtures/expected.png')
const indexPath = path.join(__dirname, '../fixtures/index.html')
const srcPath = path.join(__dirname, '../../src/index.js')

function saveScreenshot(data, path) {
    return new Promise((resolve, reject) =>
        fs.writeFile(path, data, 'base64', err =>
            (err ? reject(err) : resolve(data))
        )
    )
}

function compare(image, screenshot) {
    return new Promise((resolve) => {
        const buffer = new Buffer(screenshot, 'base64')
        resemble(image)
            .compareTo(buffer)
            .onComplete(resolve)
    })
}

function getArtifactsPath(name) {
    return path.join(__dirname, '../artifacts', name)
}

function getScript(orientation) {
    return `
        var callback = arguments[arguments.length - 1]
        var img = document.querySelector('img')

        // reset styles
        img.style['-webkit-transform'] = ''
        img.style['-webkit-transform-origin'] = ''

        var css = exif2css(${orientation})

        if (css.transform) {
            img.style['-webkit-transform'] = css.transform
        }
        if (css['transform-origin']) {
            img.style['-webkit-transform-origin'] = css['transform-origin']
        }

        img.onload = callback

        img.src = 'images/${orientation}.jpg'
    `
}

function test(orientation, expectedMismatch) {
    return driver.executeAsyncScript(getScript(orientation))
        .then(() => driver.takeScreenshot())
        .then(res => saveScreenshot(res, getArtifactsPath(`${orientation}.png`)))
        .then(res => compare(expectedImagePath, res))
        .then(res => {
            console.log(res)
            res
                .getDiffImage()
                .pack()
                .pipe(fs.createWriteStream(getArtifactsPath(`${orientation}_diff.png`)))

            assert(res.misMatchPercentage <= expectedMismatch)
        })
}

module.exports = {
    before: {
        'create bundle': () => {
            return new Promise((resolve, reject) => {
                const b = browserify([ srcPath ], { standalone: 'exif2css' })
                const stream = b.bundle((err, buf) =>
                    err ? reject(err) : resolve(String(buf))
                )
            })
                .then(res => {
                    bundle = res
                })
        },
        'init webdriver': () => {
            driver = new webdriver.Builder()
                .withCapabilities(customPhantom)
                .build()
            return driver
        },
        'go to the index page': () => {
            return driver.get(`file://${indexPath}`)
        },
        'export source code to browser': () => {
            return driver.executeScript(bundle)
        },
    },
    spec: {
        'should transform 1 correctly': () => {
            return test(1, 0)
        },
        'should transform 2 correctly': () => {
            return test(2, 1)
        },
        'should transform 3 correctly': () => {
            return test(3, 2)
        },
        'should transform 4 correctly': () => {
            return test(4, 2)
        },
        'should transform 5 correctly': () => {
            return test(5, 1)
        },
        'should transform 6 correctly': () => {
            return test(6, 1)
        },
        'should transform 7 correctly': () => {
            return test(7, 1)
        },
        'should transform 8 correctly': () => {
            return test(8, 1)
        },
    },
}


src/index.js
const propMap = {
  'r': 'rotate',
  'ry': 'rotateY',
  't': 'translateX',
  'ty': 'translateY',
}

const transformsMap = {
  '2': {
    'ry': 180,
  },
  '3': {
    'r': 180,
  },
  '4': {
    'r': 180,
    'ry': 180,
  },
  '5': {
    'r': 270,
    'ry': 180,
  },
  '6': {
    'ty': -1,
    'r': 90,
  },
  '7': {
    'ty': -1,
    't': -1,
    'r': 90,
    'ry': 180,
  },
  '8': {
    't': -1,
    'r': 270,
  },
}

const transformOriginMap = {
  '5': 'top left',
  '6': 'bottom left',
  '7': 'bottom right',
  '8': 'top right',
}

function expandTransforms(transforms) {
  const o = {}
  let expanded = false
  for (let prop in transforms) {
    if (!expanded) expanded = true
    var ep = propMap[prop]
    o[ep] = transforms[prop]
  }
  return expanded ? o : null
}

function getValue(prop, value) {
  if (prop === 'r' || prop === 'ry') {
    return `${value}deg`
  }
  if (prop === 't' || prop === 'ty') {
    return `${value * 100}%`
  }
}

function expandTransform(transforms) {
  const a = []
  for (let prop in transforms) {
    const ep = propMap[prop]
    a.push(ep + '(' + getValue(prop, transforms[prop]) + ')')
  }
  return a.length ? a.join(' ') : null
}

function expandTransformStrings(transforms) {
  const o = {}
  let expanded = false
  for (let prop in transforms) {
    if (!expanded) expanded = true
    const ep = propMap[prop]
    o[ep] = ep + '(' + getValue(prop, transforms[prop]) + ')'
  }
  return expanded ? o : null
}

/**
 * Takes the input EXIF orientation and returns the CSS rules needed to display the image correctly in the browser.
 * @param {(number|string)} orientation The EXIF orientation.
 * @returns {Exif2CssReturn} An object with `transform`, `transform-origin` (not shown in JSDoc because of hyphen), `transforms` and `transformStrings` properties.
 */
function exif2css(orientation) {
  const s = `${orientation}`
  const transforms = transformsMap[s]

  const transform = expandTransform(transforms)
  const transformOrigin = transformOriginMap[s]
  const allTransforms = expandTransforms(transforms)
  const allTransformStrings = expandTransformStrings(transforms)

  const css = {}
  if (transform) {
    css['transform'] = transform
  }
  if (transformOrigin) {
    css['transform-origin'] = transformOrigin
  }
  if (allTransforms) {
    css['transforms'] = allTransforms
  }
  if (allTransformStrings) {
    css['transformStrings'] = allTransformStrings
  }
  return css
}

/* documentary types/index.xml */
/**
 * @typedef {Object} Exif2CssReturn The return type of the function.
 * @prop {string} [transform] The complete CSS `transform` rule that contains all transforms.
 * @prop {('top left'|'top right'|'bottom left'|'bottom right')} [transform-origin] The transform origin CSS rule for orientations >= 5.
 * @prop {{translateY: number, translateX: number, rotate: number, rotateY: number}} [transforms] The raw transforms as numbers, where translates are either `-1` or `1` and rotations are either `90`, `180` and `270`.
 * @prop {{translateY: string, translateX: string, rotate: string, rotateY: string}} [transformStrings] The transforms split by individual rules that can be applied in the browser.
 */

export default exif2css